### PR TITLE
log format

### PR DESCRIPTION
**Pull request by owny990312**

### CODE REVIEW

File: src/runtime/core/logger/formatter.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  ([](const char* p, bool flag) -> const char* { \
+    if (!flag) return p;                         \
+    const char* last = strrchr(p, '/');          \
+    return last ? last + 1 : p;                  \
+  })(path, only_filename)
+
+namespace aimrt::runtime::core::logger {
+
+// check pattern
+[[nodiscard]] inline std::string Pattern(const std::string& pattern) {
```
Comment (zhangyi1357) [inline src/runtime/core/logger/formatter.h:21]: Function name like `ParsePattern" might be better?

File: src/runtime/core/logger/formatter.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    // check each part of pattern is valid
+    std::string_view pat = std::string_view(pattern).substr(i, 2);
+    if (!pattern_map.contains(pat)) {
+      throw std::runtime_error("Unknown logger pattern: " + std::string(pat));
+    }
+
+    // convert pattern to format string
+    result += "{" + std::to_string(pattern_map.at(pat)) + "}";
```
Comment (zhangyi1357) [inline src/runtime/core/logger/formatter.h:63]: Here `at` and upper `contains` both perform lookup in the hash table, which is unnecessary; we can perform the lookup only once by storing the iterator returned by `find`.

File: src/runtime/core/logger/formatter.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      {"%s", 5},  // file absolute path
+      {"%S", 5},  // todo: file name
+      {"%L", 6},  // row number
+      {"%C", 7},  // column number
+      {"%F", 8},  // function name
+      {"%v", 9}   // content
+  };
+
+  // input pattern cant be empty
+  if (pattern.empty()) [[unlikely]] {
```
Comment (zhangyi1357) [inline src/runtime/core/logger/formatter.h:38]: Empty check has been done twice.

File: src/runtime/core/logger/formatter.h [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+ public:
+  aimrt_fmt::memory_buffer Format(const LogDataWrapper& log_data_wrapper) {
+    aimrt_fmt::memory_buffer buffer;
+    for (const auto& handler : format_handlers_) {
+      handler(log_data_wrapper, buffer);
+    }
+    return buffer;
+  }
+
+  void SetPattern(std::string pattern) {
```
Comment (AimRT) [inline src/runtime/core/logger/formatter.h:27]: use const std::string&

File: src/runtime/core/logger/formatter.h [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+#include "core/logger/log_data_wrapper.h"
+#include "core/logger/log_level_tool.h"
+#include "format.h"
+#include "time_util.h"
+
+namespace aimrt::runtime::core::logger {
+
+class LogFormatter {
+ public:
+  aimrt_fmt::memory_buffer Format(const LogDataWrapper& log_data_wrapper) {
```
Comment (AimRT) [inline src/runtime/core/logger/formatter.h:19]: this is a big struct, considering the cost of copying

File: src/runtime/core/logger/formatter.h [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+#pragma once
+
+#include <cstring>
+#include <functional>
+#include <string>
+#include <vector>
+#include "core/logger/log_data_wrapper.h"
+#include "core/logger/log_level_tool.h"
+#include "format.h"
```
Comment (AimRT) [inline src/runtime/core/logger/formatter.h:12]: #include "util/format.h"

File: src/runtime/core/logger/formatter.h [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+#include "time_util.h"
```
Comment (AimRT) [inline src/runtime/core/logger/formatter.h:13]: #include "util/time_util.h"

File: src/runtime/core/logger/formatter.h [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+namespace aimrt::runtime::core::logger {
+
+class LogFormatter {
```
Comment (AimRT) [inline src/runtime/core/logger/formatter.h:17]: add unit test

File: src/runtime/core/logger/formatter_test.cc [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+// Test the GetRemappedFuncName function
+TEST(FORMATTER_TEST, Format_test) {
+  auto date = std::chrono::year(2024) / std::chrono::month(10) / std::chrono::day(1);
+  auto time = std::chrono::hours(2) +
+              std::chrono::minutes(10) +
+              std::chrono::seconds(10) +
+              std::chrono::microseconds(123456);
+
+  // // 2024-10-01 10:10:10.123456
```
Comment (AimRT) [inline src/runtime/core/logger/formatter_test.cc:17]: 2024-10-01 2:10:10.123456 ？

File: src/runtime/core/logger/console_logger_backend.cc [hunk 10]
```diff
@@ -147,18 +156,9 @@ void ConsoleLoggerBackend::Log(const LogDataWrapper& log_data_wrapper) noexcept
...
-        LogLevelTool::GetLogLevelName(log_data_wrapper.lvl),
-        log_data_wrapper.thread_id,
-        log_data_wrapper.module_name,
-        log_data_wrapper.file_name,
-        log_data_wrapper.line,
-        log_data_wrapper.column,
-        log_data_wrapper.function_name,
-        std::string_view(log_data_wrapper.log_data, log_data_wrapper.log_data_size));
+    std::string log_data_str;
+    log_data_str.reserve(log_data_wrapper.log_data_size + 256);  // reserve space for log data
```
Comment (AimRT) [inline src/runtime/core/logger/console_logger_backend.cc:160]: move to 'Format' method

**Pull request merged**